### PR TITLE
Fixed reading apiKey from nuget.config

### PR DESCRIPTION
**Pull request by zhili1208**
https://github.com/NuGet/Home/issues/2627

### CODE REVIEW

File: src/NuGet.Core/NuGet.Configuration/PackageSource/NuGetConstants.cs [hunk 1]
```diff
@@ -13,7 +13,7 @@ public static class NuGetConstants
         public static readonly string V2LegacyFeedUrl = "https://go.microsoft.com/fwlink/?LinkID=230477";
 
         public static readonly string V1FeedUrl = "https://go.microsoft.com/fwlink/?LinkID=206669";
-        public static readonly string DefaultGalleryServerUrl = "https://www.nuget.org";
+        public static readonly string DefaultGalleryServerUrl = "https://api.nuget.org/v3/index.json";
```
Comment (spadapet) [inline src/NuGet.Core/NuGet.Configuration/PackageSource/NuGetConstants.cs]: Why was this change needed?

Comment (spadapet) [inline src/NuGet.Core/NuGet.Configuration/PackageSource/NuGetConstants.cs]: Actually, now that I read the notes in the issue, I see why this change was made. I just don't know how big of an effect this URL has.

So, this looks good to me :shipit:

Comment (zhili1208) [inline src/NuGet.Core/NuGet.Configuration/PackageSource/NuGetConstants.cs]: Looks like only push and setapikey use this, www.nuget.org is not a endpoint, I don't know why we put this URL here, anyway, will double check this. @emgarten

Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Configuration/PackageSource/NuGetConstants.cs]: What is the impact of this change?

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs [hunk 2]
```diff
@@ -1494,20 +1494,18 @@ public void PushCommand_PushToServerV3_ApiKeyFromConfig()
...
-                                                        <add key='nuget.org' value='{1}' />
                                                       </packageSources>
                                                       <packageRestore>
                                                         <add key='enabled' value='True' />
                                                         <add key='automatic' value='True' />
                                                       </packageRestore>
                                                       <disabledPackageSources />
                                                       <apikeys>
-                                                        <add key='{2}' value='{3}' />
+                                                        <add key='{1}' value='{2}' />
```
Comment (yishaigalatzer) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs]: Can we validate that we can still push to
1. NuGet.org V2
2. NuGet.org V3

when both feeds are in nuget.config and when only one of them is in?

**Comment by emgarten** [comment]
The default push source was removed for push, I think it should be removed for setAPIKey also.

This fix looks like it will make things work for v3, but not v2. We would need to store the api key for both nuget v2 and nuget v3 if no source was passed in to make this work properly, but that behavior would be really odd. Let's require that the user pass the source they want to set the key for.

:watch:

**Comment by yishaigalatzer** [comment]
@emgarten lets discuss removing the default source for setApiKey. Can you schedule 15 minutes tomorrow or later today?

**Comment by alpaix** [comment]
Re-wrote the whole thing while focusing mainly on back-compat. This change adds additional API Key lookup by source url. Older mappings like 'https://www.nuget.org/api/v2/package' or 'https://www.nuget.org' will continue to work.

//cc @joelverhagen @emgarten @yishaigalatzer @rrelyea

**Comment by joelverhagen** [comment]
Looks good. :shipit:

File: src/NuGet.Core/NuGet.Commands/Utility/CommandRunnerUtility.cs [hunk 3]
```diff
@@ -41,14 +45,26 @@ public static string ResolveSymbolSource(IPackageSourceProvider sourceProvider,
             return symbolSource;
         }
 
-        public static string GetApiKey(ISettings settings, string source, string apiKey)
+        public static string GetApiKey(ISettings settings, string endpoint, string source, string defaultApiKey = null, bool isSymbols = false)
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Commands/Utility/CommandRunnerUtility.cs]: default parameters on public methods is discouraged in the nuget code

**Comment by emgarten** [comment]
:shipit:

File: src/NuGet.Core/NuGet.Commands/Utility/CommandRunnerUtility.cs (hunk 3 above)
Comment (alpaix) [inline src/NuGet.Core/NuGet.Commands/Utility/CommandRunnerUtility.cs]: Fixed
